### PR TITLE
Parse received addrv2 messages

### LINKED ISSUE #2681: ZIP-155: Parse `addrv2` in Zebra
## Motivation

[ZIP-155](https://zips.z.cash/zip-0155) will require support for parsing inbound `addrv2` messages.
But nodes can still keep sending `addr` messages.

We'll need to implement `addrv2` parsing in both Zebra (in Rust) and the DNS Seeders (in Go).

This is the ticket for Zebra.

### Scope

Zebra only needs to parse IPv4 and IPv6 in `addrv2`.

Parsing other address types is out of scope - they can be ignored.
Sending `addrv2` is out of scope.

## Deadline

ZIP-155 activates as of the NU5 mainnet activation network protocol version.
(On both mainnet and testnet.)

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

ZIP-155 activates at the same time as NU5 on mainnet.

To implement this ZIP, Zebra needs to parse received `addrv2` messages.
Sending `addrv2` messages is out of scope.

### Specifications

https://zips.z.cash/zip-0155#specification

## Solution

- add an `AddrV2` type for parsing `addrv2` messages
- update the message decoder to use `AddrV2` for parsing `addrv2` messages
- limit `addr` and `addrv2` messages to `MAX_ADDRS_IN_MESSAGE`
  - this is a bug fix for received `addr` messages
- add tests for `AddrV2`

Closes #2681.

## Review

I'd like both @jvff and @oxarbitrage to review this change, because it's the riskiest part of the ticket.

This is the last in the series of `addrv2` PRs. This PR is based on PR #3032.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

We should try to get some test vectors from the `zcashd` team.

### CODE REVIEW

File: zebra-network/src/protocol/external/addr/v2.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+///
+/// https://zips.z.cash/zip-0155#specification
+pub const ADDR_V2_IPV4_ADDR_SIZE: usize = 4;
+
+/// The size of [`Ipv6Addr`]s in `addrv2` messages.
+///
+/// https://zips.z.cash/zip-0155#specification
+pub const ADDR_V2_IPV6_ADDR_SIZE: usize = 16;
+
+/// A version 2 node address, its advertised services, and last-seen time.
```
Comment (jvff) [inline zebra-network/src/protocol/external/addr/v2.rs]: Nit: Not sure if this is useful, but I think there might be a tiny ambiguity in `a version 2 node address`:

1. An address of a version 2 node
2. A version 2 address of a node

Not sure if it's worth to reword this. If there's an easy way to change this, then maybe it's a good idea. The risk of miss-understanding is pretty low due to the context, so this could be left as is.

```suggestion
/// A version 2 address of a node, its advertised services, and last-seen time.
```

File: zebra-network/src/protocol/external/addr/v2.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            if addr.len() != ADDR_V2_IPV4_ADDR_SIZE {
+                return Err(SerializationError::Parse(
+                    "IPv4 field length did not match ADDR_V2_IPV4_ADDR_SIZE in addrv2 message",
+                ));
+            }
+
+            // > The IPV4 and IPV6 network IDs use addresses encoded in the usual way
+            // > for binary IPv4 and IPv6 addresses in network byte order (big endian).
+            let ip: [u8; ADDR_V2_IPV4_ADDR_SIZE] = addr.try_into().expect("just checked length");
+            let ip = Ipv4Addr::from(ip);
```
Comment (jvff) [inline zebra-network/src/protocol/external/addr/v2.rs]: Optional suggestion: This could be `IpAddr::from(ip)`, which would remove the need for `ip: ip.into()` in the result, with the downside of a little less explicitness.

One wilder idea would be to have some sort of generic helper function that could be reused for both IPv4 and IPv6:

```rust
fn convert_addr_bytes_to_ip_addr<const SIZE: usize>(
    addr_bytes: Vec<u8>,
) -> Result<IpAddr, SerializationError>
where
    IpAddr: From<[u8; SIZE]>,
{
    // > Clients MUST reject messages that contain addresses that have
    // > a different length than specified in this table for a specific network ID,
    // > as these are meaningless.
    if addr_bytes.len() != SIZE {
        return Err(SerializationError::Parse(
            "IPv4 field length did not match the expected IP address size in addrv2 message",
        ));
    }

    // > The IPV4 and IPV6 network IDs use addresses encoded in the usual way
    // > for binary IPv4 and IPv6 addresses in network byte order (big endian).
    let ip: [u8; SIZE] = addr_bytes.try_into().expect("just checked length");

    Ok(IpAddr::from(ip))
}
```

**Review by jvff — COMMENTED** [review]
Looks great! I added two suggestions, but they're optional.

File: zebra-network/src/protocol/external/addr/v2.rs (hunk 2 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/addr/v2.rs]: I think this change would be risky, because if Rust added extra `From` impls, Zebra would automatically accept different formats.

I don't think that's particularly likely, but I don't want to have to watch out for it, either.

File: zebra-network/src/protocol/external/addr/v2.rs (hunk 1 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/addr/v2.rs]: I agree it's confusing, what do you think of af76e24?

File: zebra-network/src/protocol/external/addr/v2.rs (hunk 2 above)
Comment (jvff) [reply zebra-network/src/protocol/external/addr/v2.rs]: I'm not sure I agree. This would be a private helper function, so Zebra wouldn't automatically accept different formats unless code was added to use it with different sizes. So the effort would be the same, or less considering that we don't have to check or review very similar code blocks to see if they are correct, we only need to see if the helper function is called correctly.

But this is optional, we can discuss this again in the future :slightly_smiling_face:

File: zebra-network/src/protocol/external/addr/in_version.rs [hunk 3]
```diff
@@ -24,8 +24,8 @@ use crate::protocol::external::arbitrary::addr_v1_ipv6_mapped_socket_addr_strate
 
 use super::{canonical_socket_addr, v1::ipv6_mapped_socket_addr};
 
-/// A version 1 node address and services, without a last-seen time.
-/// This struct is serialized and deserialized as part of `version` messages.
+/// The format used for Bitcoin node addresses in `version` messages.
+/// Contains a node address and services, without a last-seen time.
```
Comment (jvff) [inline zebra-network/src/protocol/external/addr/in_version.rs:28]: Tiny nit: I'd separate the two lines so that the documentation summary is shorter and the rest of the information is only in the detail section.

```suggestion
/// The format used for Bitcoin node addresses in `version` messages.
///
/// Contains a node address and services, without a last-seen time.
```

File: zebra-network/src/protocol/external/addr/v1.rs [hunk 4]
```diff
@@ -28,8 +28,9 @@ use proptest_derive::Arbitrary;
 #[cfg(any(test, feature = "proptest-impl"))]
 use crate::protocol::external::arbitrary::addr_v1_ipv6_mapped_socket_addr_strategy;
 
-/// A version 1 node address, its advertised services, and last-seen time.
-/// This struct is serialized and deserialized into `addr` messages.
+/// The first format used for Bitcoin node addresses.
+/// Contains a node address, its advertised services, and last-seen time.
```
Comment (jvff) [inline zebra-network/src/protocol/external/addr/v1.rs:32]: Tiny nit: same as above.

```suggestion
/// The first format used for Bitcoin node addresses.
///
/// Contains a node address, its advertised services, and last-seen time.
```

File: zebra-network/src/protocol/external/addr/v1.rs [hunk 5]
```diff
@@ -131,10 +132,8 @@ pub(in super::super) const ADDR_V1_SIZE: usize = 4 + 8 + 16 + 2;
 
 impl TrustedPreallocate for AddrV1 {
     fn max_allocation() -> u64 {
-        // Since a maximal serialized Vec<AddrV1> uses at least three bytes for its length
-        // (2MB  messages / 30B AddrV1 implies the maximal length is much greater than 253)
-        // the max allocation can never exceed (MAX_PROTOCOL_MESSAGE_LEN - 3) / META_ADDR_SIZE
-        ((MAX_PROTOCOL_MESSAGE_LEN - 3) / ADDR_V1_SIZE) as u64
+        // Since ADDR_V2_MIN_SIZE is less than 2^5, the maximum length takes up 5 bytes.
```
Comment (jvff) [inline zebra-network/src/protocol/external/addr/v1.rs]: Typo?

```suggestion
        // Since ADDR_V1_MIN_SIZE is less than 2^5, the maximum length takes up 5 bytes.
```

File: zebra-network/src/protocol/external/addr/v2.rs [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        {
+            let addr = SocketAddr::new(ip, port);
+
+            Ok(MetaAddr::new_gossiped_meta_addr(
+                addr,
+                untrusted_services,
+                untrusted_last_seen,
+            ))
+        } else {
+            Err(AddrV2UnimplementedError)
```
Comment (jvff) [inline zebra-network/src/protocol/external/addr/v2.rs]: Nit: The name of this error could be ambiguous. Maybe it should be called `UnsupportedAddrV2TypeError` or `UnsupportedAddrV2VariantError`?

File: zebra-network/src/protocol/external/addr/v2.rs [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+                    writer.write_u8(ADDR_V2_IPV6_NETWORK_ID)?;
+
+                    let ip: [u8; ADDR_V2_IPV6_ADDR_SIZE] = ip.octets();
+                    zcash_serialize_bytes(&ip.to_vec(), &mut writer)?;
+
+                    writer.write_u16::<BigEndian>(*port)?;
+                }
+            }
+        } else {
+            panic!("unexpected AddrV2 variant: {:?}", self);
```
Comment (jvff) [inline zebra-network/src/protocol/external/addr/v2.rs]: Nit: maybe use `unreachable!` for a tiny bit more semantics?

```suggestion
            unreachable!("unexpected AddrV2 variant: {:?}", self);
```

File: zebra-network/src/protocol/external/tests/preallocate.rs [hunk 8]
```diff
@@ -117,3 +117,53 @@ proptest! {
...
+        ) = max_allocation_is_big_enough(addr);
+
+        // Check that our smallest_disallowed_vec is only one item larger than the limit
+        prop_assert!(((smallest_disallowed_vec_len - 1) as u64) == AddrV2::max_allocation());
+        // Check that our smallest_disallowed_vec is too big to send in a valid Zcash message
+        prop_assert!(smallest_disallowed_serialized_len > MAX_PROTOCOL_MESSAGE_LEN);
+
+        // Check that our largest_allowed_vec contains the maximum number of AddrV2s
+        prop_assert!((largest_allowed_vec_len as u64) == AddrV2::max_allocation());
+        // largest_allowed_serialized_len exceeds the limit for variable-sized types
```
Comment (jvff) [inline zebra-network/src/protocol/external/tests/preallocate.rs]: Question: I don't understand this comment, what does it mean? :sweat_smile:

File: zebra-network/src/protocol/external/tests/prop.rs [hunk 9]
```diff
@@ -227,4 +233,139 @@ proptest! {
...
+            gossiped_addr,
+            hex::encode(&addr_bytes),
+            deserialized_addr,
+            hex::encode(&addr_bytes2),
+        );
+    }
+
+    /// Test round-trip AddrV2 serialization for all MetaAddr variants after sanitization
+    #[test]
+    fn addr_v2_sanitized_roundtrip(addr in any::<MetaAddr>()) {
```
Comment (jvff) [inline zebra-network/src/protocol/external/tests/prop.rs:179]: This test seems like a superset of the previous test. Is there a reason to keep both?

**Comment by zancas** [comment]
OK, I've added a few tests to a `zcashd` feature branch.   The test vectors are composed of addresses and serializations that were already in the tests elsewhere, but they are conveniently adjacent here:

https://github.com/zingolabs/zcash/blob/9ee66e423a3fbf4829ffeec354e82f4fbceff864/src/test/netbase_tests.cpp#L481

**Review by jvff — DISMISSED** [review]
Overall looks good. There's one documentation typo, and I also added some optional suggestions.

File: zebra-network/src/protocol/external/addr/in_version.rs (hunk 3 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/addr/in_version.rs:28]: The different contents of each format are actually really important, so I'd like them to appear in the summary.
(And they're also poorly documented in Bitcoin.)

File: zebra-network/src/protocol/external/addr/v2.rs (hunk 7 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/addr/v2.rs]: Technically that code is reachable by mistake, particularly from proptests.
But you're right, it's intended to be unreachable.

Fixed in c2653838.

File: zebra-network/src/protocol/external/addr/v1.rs (hunk 5 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/addr/v1.rs]: Fixed and clarified in a7de5a2cc.

File: zebra-network/src/protocol/external/addr/v2.rs (hunk 6 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/addr/v2.rs]: Fixed in 6dd721243 and caa0daedf, along with the corresponding variant name and comments.

File: zebra-network/src/protocol/external/tests/preallocate.rs (hunk 8 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/tests/preallocate.rs]: It explains why we skip checking `largest_allowed_serialized_len` for variable-sized types.

Clarified in 337f95374.

File: zebra-network/src/protocol/external/tests/prop.rs (hunk 9 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/tests/prop.rs:179]: They used to be different, but they're the same now.

Deleted the redundant tests in bba2dade6.

File: zebra-network/src/protocol/external/addr/v2.rs (hunk 2 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/addr/v2.rs]: Now I understand how you wanted the new method to be used!

See 50ba6526d.

**Comment by teor2345** [comment]
I'm going to put this in draft until I add test vectors from `zcashd`, and test vectors with extreme service values and address lengths.

**Comment by teor2345** [comment]
> OK, I've added a few tests to a `zcashd` feature branch. The test vectors are composed of addresses and serializations that were already in the tests elsewhere, but they are conveniently adjacent here:
> 
> https://github.com/zingolabs/zcash/blob/9ee66e423a3fbf4829ffeec354e82f4fbceff864/src/test/netbase_tests.cpp#L481

@zancas thanks, we're now testing these cases in Zebra as well.

I added some extra addrv1 and addrv2 cases to cover various parsing edge-cases.
You can see them at:
https://github.com/ZcashFoundation/zebra/blob/f246cbc80e0e20917779dced46dae884f26d48cd/zebra-test/src/network_addr.rs#L46
https://github.com/ZcashFoundation/zebra/blob/f246cbc80e0e20917779dced46dae884f26d48cd/zebra-test/src/network_addr.rs#L122

Our test implementation is here, but it's probably not that interesting for you:

https://github.com/ZcashFoundation/zebra/blob/f246cbc80e0e20917779dced46dae884f26d48cd/zebra-network/src/protocol/external/tests/vectors.rs#L39

File: zebra-test/src/network_addr.rs [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+        // IPv4
+        <Vec<u8>>::from_hex(
+            concat!(
+                "01", // number of entries
+
+                "61bc6649",                         // time, Fri Jan  9 02:54:25 UTC 2009
+                "0100000000000000",                 // service flags, NODE_NETWORK
+                "00000000000000000000ffff",         // address, fixed 16 bytes (IPv4 embedded in IPv6)
+                "7f000001",                         // IPv4, 127.0.0.1
```
Comment (jvff) [inline zebra-test/src/network_addr.rs]: Nit: minor tweak for clarity:

```suggestion
                "00000000000000000000ffff",         // address, first 12 bytes of fixed 16 bytes
                                                    // (prefix for IPv4 embedded in IPv6)
                "7f000001",                         // final 4 bytes for IPv4, 127.0.0.1
```

File: zebra-test/src/network_addr.rs [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+                "02",                               // network id, IPv6
+                "10",                               // address length, COMPACTSIZE(16)
+                "00000000000000000000000000000001", // address
+                "00f1",                             // port
+
+                "79627683",                         // time, Tue Nov 22 11:22:33 UTC 2039
+                "01",                               // service flags, COMPACTSIZE(NODE_NETWORK)
+                "04",                               // network id, TorV3
+                "20",                               // address length, COMPACTSIZE(32)
+                "53cd5648488c4707914182655b7664034e09e66f7e8cbf1084e654eb56c5bd88", // 32 bytes of key + 2 bytes checksum
```
Comment (jvff) [inline zebra-test/src/network_addr.rs]: Question: should this be `30 bytes of key + 2 bytes of checksum`?

File: zebra-network/src/protocol/external/tests/vectors.rs [hunk 12]
```diff
@@ -24,3 +35,254 @@ fn parses_msg_wtx_inventory_type() {
...
+    let codec = Codec::builder().finish();
+
+    for (case_idx, addr_v1_bytes) in zebra_test::network_addr::ADDR_V1_IP_VECTORS
+        .iter()
+        .enumerate()
+    {
+        let mut addr_v1_bytes = io::Cursor::new(addr_v1_bytes);
+
+        let deserialized: Message = codec
+            .read_addr(&mut addr_v1_bytes)
```
Comment (jvff) [inline zebra-network/src/protocol/external/tests/vectors.rs]: Minor: I think you can avoid the `Cursor` if you want to:

```suggestion
        let deserialized: Message = codec
            .read_addr(&mut addr_v1_bytes.as_slice())
```

**Review by jvff — DISMISSED** [review]
Looks good! I added a few suggestions, but they are all optional :+1:

File: zebra-test/src/network_addr.rs (hunk 11 above)
Comment (teor2345) [reply zebra-test/src/network_addr.rs]: No, the checksum only exists to avoid mistakes in human address entry. It is deterministically calculated from the public key, so it doesn't need to be sent by machines.

(I helped design this format!)

Fixed in a7f11c210.

File: zebra-test/src/network_addr.rs (hunk 10 above)
Comment (teor2345) [reply zebra-test/src/network_addr.rs]: Fixed in a7f11c210

File: zebra-network/src/protocol/external/tests/vectors.rs (hunk 12 above)
Comment (teor2345) [reply zebra-network/src/protocol/external/tests/vectors.rs]: Fixed in 96e6121aa

**Comment by teor2345** [comment]
I applied the optional suggestions.

@zancas commit a7f11c2 fixes a mistake in the comments about the Tor v3 address test vectors. (I also listed all the field values in those comments.)

**Review by jvff — APPROVED** [review]

File: zebra-test/src/network_addr.rs (hunk 11 above)
Comment (jvff) [reply zebra-test/src/network_addr.rs]: > (I helped design this format!)

That's awesome! :star_struck:
